Tests/test_file_apng.py

```diff
@@ -663,6 +665,36 @@ def test_apng_save_blend(tmp_path):
         assert im.getpixel((0, 0)) == (0, 255, 0, 255)
 
 
+def test_save_all_progress():
+    out = BytesIO()
+    progress = []
+
+    def callback(filename, frame_number, n_frames):
+        progress.append((filename, frame_number, n_frames))
+
+    Image.new("RGB", (1, 1)).save(out, "PNG", save_all=True, progress=callback)
+    assert progress == [(0, 1, 1)]
+
+    out = BytesIO()
+    progress = []
+
+    with Image.open("Tests/images/apng/single_frame.png") as im:
+        with Image.open("Tests/images/apng/delay.png") as im2:
+            im.save(
+                out, "PNG", save_all=True, append_images=[im, im2], progress=callback
+            )
+
+    assert progress == [
+        (0, 1, 7),
+        (1, 2, 7),
+        (2, 3, 7),
```

Comment:
Say... if the frames are counted from `1`, why count the images from `0`?

Reply:
I'm triggering these callbacks when frames are completed, not when they are started.

My intention is for `(0, 1, 7)` to communicate "We're in the first image. We've completed the first frame, there are seven frames in total".
Then `(2, 7, 7)` communicates "We're in the third image. We've completed seven frames, there are seven frames in total, so we're done"

If I started counting the frames from 0, the last value would be `(2, 6, 7)`. That looks less to me like save_all is finished.

I recognise this could be a subject of debate.

Reply:
I was trying to explain that my goal is not to number the frames from 1, but rather to clearly communicate when progress is completed, because the frame count will be compared with the total number of frames. The image count will not.

> arrays in C are indexed starting from 0, but that's an implementation detail and it's kind of ridiculous to drag it into higher-level languages

Python arrays are also indexed from 0.

Reply:
> The image count will not.

And yet counting from `0` would be inconsistent with the other counting used _within same data set_.

> Python arrays are also indexed from 0.

Which makes no sense since Python is a high-level language (and there's some consistency issues caused by this arbitrary choice to imitate C array indexing, but that's neither here nor there).

Besides, the images aren't coming from a list as far as the user is concerned (well, there's `append_images`, but going by that logic you'd have to start from `-1` since the first input image is not in this list in the first place… except that `-1` is a valid index in Python, so it probably would be `None` instead).

Reply:
I've pushed a commit naming each piece of information within the returned data. It is now "image_index" and "completed_frames". Hopefully that clarifies that one is the Pythonic zero-based index of the image in progress, and the other is the number of frames that have been finished.